File: frontend/src/lib/models/page_elements/button_input_element_config.ts
import type { PageElementConfig } from "../page_element_config";

export let BUTTON_ELEMENT_TYPE = "button";

export interface ButtonInputElementConfig extends PageElementConfig {
    label: string;
    options?: ButtonInputOption;
    action: ButtonAction;
}

export interface ButtonInputOption {
    color?: string;
    width?: string;
    height?: string;
}

export interface ButtonAction {
    type: string;
}

export let BUTTON_ACTION_TYPE_SUBMIT = "submit";

export interface ButtonSubmitAction extends ButtonAction {
    endpoint: string;
    method?: string;
    values?: ButtonSubmitActionValue[];
}

export interface ButtonSubmitActionValue {
    key: string;
    type: string;
}

export let BUTTON_ACTION_VALUE_VARIABLE = "var";

export interface VariableButtonSubmitActionValue extends ButtonSubmitActionValue {
    variable: string;
}

export let BUTTON_ACTION_VALUE_BINDING = "binding";

export interface BindingButtonSubmitActionValue extends ButtonSubmitActionValue {
    identifier: string;
}

File: src/main/kotlin/dev/schlaubi/dashboard/models/pageelements/ButtonInputElementConfig.kt
package dev.schlaubi.dashboard.models.pageelements

import dev.schlaubi.dashboard.models.ActionValue
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable


@Serializable
@SerialName("button")
data class ButtonInputElementConfig(
    val label: String, val options: ButtonInputOption?, val action: ButtonAction
) : PageElementConfig()

@Serializable
data class ButtonInputOption(
    val color: String? = null,
    val width: String? = null,
    val height: String? = null,
)

@Serializable
sealed class ButtonAction

@Serializable
@SerialName("submit")
data class ButtonSubmitAction(
    val endpoint: String, val method: String? = null, val values: List<ActionValue>? = null
) : ButtonAction()


